source/text/Format.cpp

```diff
@@ -389,3 +425,112 @@ vector<string> Format::Split(const string &str, const string &separator)
 	}
 	return result;
 }
+
+
+
+string Format::ExpandConditions(const string &source, ConditionGetter getter)
+{
+	// Optimization for most common case: no conditions
+	if(source.find('&') == string::npos)
+		return source;
+
+	string result;
+	result.reserve(source.size());
+
+	size_t formatStart = string::npos;
+	size_t formatSize = string::npos;
+	size_t conditionStart = string::npos;
+	size_t conditionSize = string::npos;
+
+	// Hand-coded regular grammar parser for:
+	//	&[format@condition]
+	//	&[condition]
+	// Using these states:
+	//	state = _ ----- outside of all &[] regions
+	//	state = & ----- just read & and hoping to see a [
+	//	state = [ ----- read &[ but haven't seen @ or ] yet
+	//	state = N ----- inside a nested [] of depth `depth`
+	//	state = @ ----- read &[...@ but haven't seen ] yet. Have format start & size.
+
+	// Anything inside a &[...] is sent to AppendCondition
+
+	static const char OUTER = '_';
+	static const char PREFIX = '&';
+	static const char LPAREN = '[';
+	static const char RPAREN = ']';
+	static const char DIVIDER = '@';
+	static const char NESTED = 'N';
+
+	char state = OUTER;
+	// Depth of nested [] within the &[...]
+	int depth = 0;
+	// State before entering the nested []
+	char oldState = LPAREN;
+	// "start" is the beginning of the text that has not yet been sent to result.
+	size_t start = 0;
+	for(size_t look = 0; look < source.size(); ++look)
+	{
+		char next = source[look];
+		// This would be faster with a nested select, but that would be
+		// harder to read, and I don't expect this to be performance-critical.
+		if(state == OUTER && next == PREFIX)
+			state = PREFIX;
+		else if(state == OUTER || (state == PREFIX && next != LPAREN))
+		{
+			result.append(source, start, look - start + 1);
+			start = look + 1;
+			state = OUTER;
+		}
```

Comment:
Oh, wait, there was one more thing.

In most cases, this is going to copy the string one character at a time. I see you reserve extra space in the result string to avoid constant reallocation, but is this really the best way to do this?
How would only appending and updating the `start` value when you get to a prefix (and at the end as you already have it doing)?

Reply:
The deed is done.